functions/auth.js

```diff
@@ -19,3 +30,164 @@ exports.passwordsMatch = (submittedPassword, actualPassword) =>
   submittedPassword.trim() === actualPassword.trim() ||
   lowercaseFirstChar(submittedPassword.trim()) ===
     lowercaseFirstChar(actualPassword.trim());
+
+/**
+ * Redirect the user to the authentication consent screen. The 'state' cookie is set for later state verification.
+ *
+ * @see https://github.com/lelylan/simple-oauth2/blob/3.x/API.md#authorizeurlauthorizeoptions--string
+ */
+// TODO: rename this authorize or similar?
+exports.connectI4AOAuth = functions.https.onRequest(async (req, res) => {
+  const { venueId } = req.query;
+
+  if (!AUTH_ORIGIN) {
+    throw new HttpsError("internal", "auth.origin is not configured properly");
+  }
+
+  if (!checkIfValidVenueId(venueId)) {
+    throw new HttpsError("invalid-argument", "venueId is not a valid venue id");
+  }
+
+  const authConfig = await fetchAuthConfig(venueId);
+
+  const authClient = createOAuth2Client(authConfig);
+
+  // TODO: configure this in cloud config and/or firestore or similar
+  const authCodeReturnUri = `${AUTH_ORIGIN}/auth/connect/i4a/handler?venueId=${venueId}`;
+
+  // TODO: configure this in cloud config and/or firestore or similar
+  const scope = "";
+
+  const redirectUri = authClient.authorizationCode.authorizeURL({
+    redirect_uri: authCodeReturnUri,
+    scope,
+  });
+
+  functions.logger.log("Redirecting to:", redirectUri);
+  res.redirect(redirectUri);
+});
+
+/**
+ * Exchanges a given auth code passed in the 'code' URL query parameter for a Firebase auth token.
+ * The request also needs to specify a 'state' query parameter which will be checked against the 'state' cookie.
+ */
+exports.connectI4AOAuthHandler = functions.https.onRequest(async (req, res) => {
+  const { venueId } = req.query;
+
+  if (!AUTH_ORIGIN) {
+    throw new HttpsError("internal", "auth.origin is not configured properly");
+  }
+
+  if (!checkIfValidVenueId(venueId)) {
+    throw new HttpsError("invalid-argument", "venueId is not a valid venue id");
+  }
+
+  const authConfig = await fetchAuthConfig(venueId);
+
+  const {
+    i4aApiKey,
+    i4aOAuthUserInfoUrl,
+    i4aGetUserMeetingInfoUrl,
+  } = authConfig;
+
+  const authClient = createOAuth2Client(authConfig);
+
+  const { code: authCode } = req.query;
+
+  functions.logger.log("Received auth code:", authCode);
+  const results = await authClient.authorizationCode.getToken({
+    code: authCode,
+    redirect_uri: "http://localhost:3000/in/foo", // TODO: what do we want to use here?
```

Comment:
[TODO] What should we be using for the redirect URL here? I'm not even sure if/what this is used for in this I4A flow. I think in a normal OAuth flow it might be used to check the redirect URL against a whitelist or similar?